File: documentation/manual/working/scalaGuide/advanced/routing/code/scalaguide/binder/models/AgeRange.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package scalaguide.binder.models

import scala.Left
import scala.Right

import play.api.mvc.PathBindable
import play.api.mvc.QueryStringBindable
import play.Logger

//#declaration
case class AgeRange(from: Int, to: Int) {}
//#declaration
object AgeRange {
  // #bind
  implicit def queryStringBindable(implicit intBinder: QueryStringBindable[Int]): QueryStringBindable[AgeRange] =
    new QueryStringBindable[AgeRange] {
      override def bind(key: String, params: Map[String, Seq[String]]): Option[Either[String, AgeRange]] = {
        for {
          from <- intBinder.bind("from", params)
          to   <- intBinder.bind("to", params)
        } yield {
          (from, to) match {
            case (Right(from), Right(to)) => Right(AgeRange(from, to))
            case _                        => Left("Unable to bind an AgeRange")
          }
        }
      }
      override def unbind(key: String, ageRange: AgeRange): String = {
        intBinder.unbind("from", ageRange.from) + "&" + intBinder.unbind("to", ageRange.to)
      }
    }
  // #bind
}

File: documentation/manual/working/javaGuide/advanced/routing/code/javaguide/binder/models/AgeRange.java
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package javaguide.binder.models;

import java.util.Map;
import java.util.Optional;
import play.libs.F.*;
import play.mvc.QueryStringBindable;

// import java.util.Optional;

// #declaration
public class AgeRange implements QueryStringBindable<AgeRange> {

  public Integer from;
  public Integer to;

  // #declaration

  // #bind
  @Override
  public Optional<AgeRange> bind(String key, Map<String, String[]> data) {

    try {
      from = Integer.valueOf(data.get("from")[0]);
      to = Integer.valueOf(data.get("to")[0]);
      return Optional.of(this);

    } catch (Exception e) { // no parameter match return None
      return Optional.empty();
    }
  }

  @Override
  public String unbind(String key) {
    return new StringBuilder().append("from=").append(from).append("&to=").append(to).toString();
  }

  // #bind

  @Override
  public String javascriptUnbind() {
    return new StringBuilder().append("from=").append(from).append(";to=").append(to).toString();
  }
}
